src/box/vy_run.c

```diff
@@ -1021,47 +1117,57 @@ vy_page_no_info_unref(struct vy_page *page)
 {
 	assert(page != NULL);
 	assert(page->ri_refs > 0);
+
 	page->ri_refs -= 1;
 	if (page->ri_refs == 0)
 		vy_page_delete(page);
 }
 
+/**
+ * Decrease memory-stored page's reference counter and delete it if
+ * necessary, bypassing including it in the LRU list. Used to
+ * accelerate the engine in 'no-page-cache' mode.
+ */
+static inline void
+vy_page_no_cache_unref(struct vy_page_info *page_info)		//temp/ can be replaced with no_info_unref
+{
+	struct vy_page *page = page_info->page;
+	assert(page != NULL);
+	assert(page->ri_refs > 0);
+
+	page->ri_refs -= 1;
+	if (page->ri_refs == 0) {
+		vy_page_delete(page);
+		page_info->page = NULL;
+	}
+}
+
 /**
  * Read a page from disk given its number.
- * The function caches two most recently read pages.
  *
  * @retval 0 success
  * @retval -1 critical error
  */
 static NODISCARD int
 vy_run_iterator_load_page(struct vy_run_iterator *itr, uint32_t page_no,
-			  struct vy_entry key, enum iterator_type iterator_type,
+			  struct vy_entry key,
+			  enum iterator_type iterator_type,
 			  struct vy_page **result, uint32_t *pos_in_page,
 			  bool *equal_found)
 {
 	struct vy_run *run = itr->slice->run;
-	struct vy_run_env *run_env = run->env;
+	struct vy_run_env *run_env = run->run_env;
 	struct vy_page_info *page_info = vy_run_page_info(run, page_no);
 	struct vy_page *page = page_info->page;
 
 	if (page != NULL) {
 		/* Page is already stored in memory */
+		vy_page_ref(run, page_info);
```

Comment:
This was done at the beginning of the function to avoid deleting/placing to LRU the old iterator's page in case it has refs == 1 but it is the one that is requested in this iteration.  Anyways, `vy_run_iterator_load_page()` is already completely refactored, see the last commits